src/plugins/ipt/ipt.cpp

```diff
@@ -0,0 +1,469 @@
+
+    drakvuf_release_vmi(drakvuf);
+
+    plugin->destroy_trap(info->trap);
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+static event_response_t mm_access_fault_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    addr_t fault_va = drakvuf_get_function_argument(drakvuf, info, 2);
+    PRINT_DEBUG("[IPT] MmAccessFault(%d, %lx)\n", info->proc_data.pid, fault_va);
+
+    if (fault_va & (1ULL << 63))
+    {
+        PRINT_DEBUG("[IPT] Don't trap in kernel %d %lx\n", info->proc_data.pid, fault_va);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    auto plugin = get_trap_plugin<ipt>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto trap = plugin->register_trap<access_fault_result_t>(
+                    info,
+                    mm_access_fault_return_hook_cb,
+                    breakpoint_by_pid_searcher());
+    if (!trap)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto params = get_trap_params<access_fault_result_t>(trap);
+    params->set_result_call_params(info, drakvuf_get_current_thread(drakvuf, info));
+    params->fault_va = fault_va;
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+event_response_t ipt_cr3_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    uint32_t payloads[4] =
+    {
+        PTW_CURRENT_CR3, (uint32_t) info->regs->cr3,
+        PTW_CURRENT_TID, info->proc_data.tid
+    };
+
+    annotate_ipt(drakvuf, (ipt*)info->trap->data, info->vcpu, payloads, 2);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+event_response_t ipt_catchall_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    uint32_t payloads[2] =
+    {
+        PTW_EVENT_ID, (uint32_t) info->event_uid
+    };
+
+    annotate_ipt(drakvuf, (ipt*)info->trap->data, info->vcpu, payloads, 1);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+ipt::ipt(drakvuf_t drakvuf, const ipt_config* c, output_format_t output)
+    : pluginex(drakvuf, output)
+{
+    for (unsigned int i = 0; i < this->num_vcpus; i++)
+    {
+        this->vcpu[i].fd = 0;
+    }
+
+    this->ipt_dir = c->ipt_dir;
+
+    if (!this->ipt_dir)
+    {
+        PRINT_DEBUG("[IPT] Target directory not provided, not activating IPT plugin\n");
+        return;
+    }
+
+    {
+        std::stringstream ss;
+        ss << this->ipt_dir;
+        ss << "/frames";
+
+        std::string frame_dir = ss.str();
+
+        int res = mkdir(frame_dir.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+
+        if (res != 0 && errno != EEXIST)
+        {
+            PRINT_DEBUG("[IPT] Failed to create %s\n", frame_dir.c_str());
+            throw -1;
+        }
+    }
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    this->num_vcpus = vmi_get_num_vcpus(vmi);
+    drakvuf_release_vmi(drakvuf);
+
+    if (IPT_MAX_VCPUS < num_vcpus)
+    {
+        PRINT_DEBUG("[IPT] Too many vCPUs, the excess ones will be not monitored.\n");
+        num_vcpus = IPT_MAX_VCPUS;
+    }
+
+    for (unsigned int i = 0; i < num_vcpus; i++)
```

Comment:
```suggestion
    for (auto& vcpu : this->vcpus)
```
(this requres additional changes to following for loop)

Reply:
we need to know the explicit index here, because `drakvuf_enable_ipt` needs to know for which vCPU ID we want to enable IPT feature